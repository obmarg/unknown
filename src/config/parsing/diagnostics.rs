//! Some diagnostics machinery built on miette

use std::{error::Error, fmt::Display};

use miette::{Diagnostic, SourceCode};

pub struct DynDiagnostic {
    inner: Box<dyn Diagnostic + Send + Sync + 'static>,
    source_code: Option<Box<dyn miette::SourceCode + 'static>>,
}

impl std::fmt::Debug for DynDiagnostic {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("DynDiagnostic")
            .field("inner", &self.inner)
            .finish_non_exhaustive()
    }
}

impl Display for DynDiagnostic {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.inner.fmt(f)
    }
}

impl Error for DynDiagnostic {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        self.inner.source()
    }
}

impl Diagnostic for DynDiagnostic {
    fn code<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.inner.code()
    }

    fn severity(&self) -> Option<miette::Severity> {
        self.inner.severity()
    }

    fn help<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.inner.help()
    }

    fn url<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
        self.inner.url()
    }

    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
        self.source_code
            .as_ref()
            .map(|s| s.as_ref())
            .or_else(|| self.inner.source_code())
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
        self.inner.labels()
    }

    fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn Diagnostic> + 'a>> {
        self.inner.related()
    }
}

impl DynDiagnostic {
    pub fn new(diagnostic: impl Diagnostic + Send + Sync + 'static) -> Self {
        DynDiagnostic {
            inner: diagnostic.into(),
            source_code: None,
        }
    }

    pub fn with_source_code(mut self, source_code: impl SourceCode + 'static) -> Self {
        // If inner already has source_code we trust that is correct and leave it alone
        if let None = self.inner.source_code() {
            self.source_code = Some(Box::new(source_code));
        }
        self
    }
}
