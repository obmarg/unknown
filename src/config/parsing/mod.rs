use camino::{Utf8Path, Utf8PathBuf};

mod project;
mod tasks;
mod validation;
mod workspace;

pub(super) use self::{project::ProjectDefinition, tasks::*, workspace::WorkspaceDefinition};

pub use validation::Validator;

use super::ConfigSource;

#[derive(Debug, thiserror::Error)]
#[error("Error parsing {1}")]
pub struct ParsingError(pub(super) knuffel::Error, pub(super) Utf8PathBuf);

impl miette::Diagnostic for ParsingError {
    fn code<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
        self.0.code()
    }

    fn severity(&self) -> Option<miette::Severity> {
        self.0.severity()
    }

    fn help<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
        self.0.help()
    }

    fn url<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
        self.0.url()
    }

    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
        self.0.source_code()
    }

    fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
        self.0.labels()
    }

    fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn miette::Diagnostic> + 'a>> {
        self.0.related()
    }

    fn diagnostic_source(&self) -> Option<&dyn miette::Diagnostic> {
        self.0.diagnostic_source()
    }
}

pub fn parse_project_file(config: &ConfigSource) -> Result<ProjectDefinition, ParsingError> {
    let filename = config.filename();
    let config = knuffel::parse::<ProjectDefinition>(filename, config.contents())
        .map_err(|e| ParsingError(e, filename.to_owned().into()))?;

    // TODO: At some point want to validate the data in the file.
    // e.g. names can't have commas or slashes in them etc.

    Ok(config)
}

pub fn parse_workspace_file(
    path: &Utf8Path,
    contents: &str,
) -> Result<WorkspaceDefinition, ParsingError> {
    let config = knuffel::parse(
        path.file_name()
            .expect("workspace file path to have a filename"),
        contents,
    )
    .map_err(|e| {
        let filename = Utf8PathBuf::from(path.file_name().unwrap());
        ParsingError(e, filename)
    })?;

    // TODO: At some point want to validate the data in the file.
    // e.g. names can't have commas or slashes in them etc.

    Ok(config)
}

pub fn parse_task_file(path: &Utf8Path, contents: &str) -> Result<TaskBlock, ParsingError> {
    let config = knuffel::parse(
        path.file_name()
            .expect("workspace file path to have a filename"),
        contents,
    )
    .map_err(|e| ParsingError(e, path.to_owned()))?;

    // TODO: At some point want to validate the data in the file.
    // e.g. names can't have commas or slashes in them etc.

    Ok(config)
}
