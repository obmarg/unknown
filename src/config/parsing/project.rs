use crate::config::spanned::Spanned;

use super::{super::paths::ConfigPath, tasks};

#[derive(knuffel::Decode, Debug)]
pub struct ProjectDefinition {
    #[knuffel(child, unwrap(argument))]
    pub(super) project: String,

    #[knuffel(child, default)]
    pub(in crate::config) dependencies: DependencyBlock,

    #[knuffel(child, default)]
    pub(super) tasks: tasks::TaskBlock,
}

#[derive(knuffel::Decode, Debug, Default)]
pub struct DependencyBlock {
    #[knuffel(children(name = "project"), unwrap(argument))]
    pub(in crate::config) projects: Vec<Spanned<ConfigPath>>,
}
