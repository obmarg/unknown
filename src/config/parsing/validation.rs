use crate::{
    config::{
        parsing, paths::ConfigPathValidationError, spanned::WithSpan, validated, ConfigSource,
        UnvalidatedConfig, UnvalidatedProjectFile, UnvalidatedWorkspaceFile, ValidConfig,
        ValidPath, ValidProjectFile, WorkspaceFile, WorkspaceRoot,
    },
    diagnostics::{CollectResults, ConfigError, DynDiagnostic},
};

pub struct Validator {
    workspace_root: WorkspaceRoot,
    errors: Vec<DynDiagnostic>,
}

impl Validator {
    pub fn new(workspace_root: WorkspaceRoot) -> Self {
        Validator {
            workspace_root,
            errors: vec![],
        }
    }

    pub fn ok(&mut self) -> Result<(), ConfigError> {
        if self.errors.is_empty() {
            return Ok(());
        }
        Err(ConfigError {
            errors: std::mem::take(&mut self.errors),
        })
    }

    pub fn validate_config(
        &mut self,
        config: UnvalidatedConfig,
    ) -> Result<ValidConfig, ConfigError> {
        let workspace_file = self.validate_workspace_file(config.workspace_file);

        let project_files = config
            .project_files
            .into_iter()
            .map(|project_file| self.validate_project_file(project_file))
            .collect::<Option<Vec<_>>>();

        self.ok()?;

        let (workspace_file, project_files) = workspace_file
            .zip(project_files)
            .expect("validation errors if theres any nones here");

        Ok(ValidConfig {
            workspace_file,
            project_files,
        })
    }

    fn validate_workspace_file(
        &mut self,
        workspace: UnvalidatedWorkspaceFile,
    ) -> Option<WorkspaceFile> {
        Some(WorkspaceFile {
            workspace_root: workspace.workspace_root,
            config: validated::WorkspaceDefinition {
                name: workspace.config.name,
                project_paths: workspace.config.project_paths,
            },
            source: workspace.source,
        })
    }

    fn validate_project_file(&mut self, file: UnvalidatedProjectFile) -> Option<ValidProjectFile> {
        let config =
            self.validate_project_definition(file.config, &file.project_root, &file.source)?;

        Some(ValidProjectFile {
            project_root: file.project_root,
            config,
            source: file.source,
        })
    }

    pub fn validate_project_definition(
        &mut self,
        project: parsing::ProjectDefinition,
        project_path: &ValidPath,
        config_source: &ConfigSource,
    ) -> Option<validated::ProjectDefinition> {
        let dependencies = project
            .dependencies
            .projects
            .into_iter()
            .map(|p| {
                let span = p.span;
                Ok::<_, ConfigPathValidationError>(
                    p.into_inner()
                        .validate_relative_to(project_path)?
                        .with_span(span),
                )
            })
            .collect_results();

        let tasks = self.validate_tasks(project.tasks, project_path, config_source);

        let (dependencies, tasks) = self.record_errors(dependencies, config_source).zip(tasks)?;

        Some(validated::ProjectDefinition {
            project: project.project,
            dependencies,
            tasks,
        })
    }

    pub fn validate_tasks(
        &mut self,
        tasks: parsing::TaskBlock,
        relative_to: &ValidPath,
        source_code: &ConfigSource,
    ) -> Option<validated::TaskBlock> {
        let imports = tasks
            .imports
            .into_iter()
            .map(|path| path.validate_relative_to(relative_to))
            .collect_results();

        let imports = self.record_errors(imports, source_code);

        let tasks = tasks
            .tasks
            .into_iter()
            .map(|task| self.validate_task(task, source_code))
            .collect::<Option<Vec<_>>>();

        let (imports, tasks) = imports.zip(tasks)?;

        Some(validated::TaskBlock { imports, tasks })
    }

    fn validate_task(
        &mut self,
        task: parsing::TaskDefinition,
        config_source: &ConfigSource,
    ) -> Option<validated::TaskDefinition> {
        let requires = task
            .requires
            .into_iter()
            .map(|r| r.parse(&self.workspace_root))
            .collect_results();

        let requires = self.record_errors(requires, config_source)?;

        Some(validated::TaskDefinition {
            name: task.name,
            commands: task.commands,
            requires,
            input_blocks: task.input_blocks.into_iter().map(Into::into).collect(),
            source: config_source.clone(),
        })
    }

    fn record_errors<T, E>(
        &mut self,
        res: Result<T, Vec<E>>,
        config_source: &ConfigSource,
    ) -> Option<T>
    where
        E: miette::Diagnostic + Send + Sync + 'static,
    {
        match res {
            Ok(inner) => Some(inner),
            Err(errors) => {
                for error in errors {
                    self.errors
                        .push(DynDiagnostic::new(error).with_source_code(config_source.clone()));
                }
                None
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use camino::Utf8PathBuf;

    use super::*;
    use crate::test_files::TestFiles;

    #[test]
    fn project_validation_happy_path() {
        let files = TestFiles::new()
            .with_file("library/project.kdl", r#"project "library"#)
            .with_file("service/project.kdl", "");
        let source = ConfigSource::new(
            "service/project.kdl",
            r#"
                project "service"
                dependencies {
                    project "/library"
                    project "../library"
                }

                tasks {
                    task "build" {
                        // These technically aren't valid, but the ways they're wrong
                        // are validated later on in the process.
                        requires "a-task-in-library" in="library"
                        requires "a-task-in-our-deps" in="^self"
                        requires "a-task-in-ourselves" in="self"
                        requires "a-task-without-an-in-specified"

                        command "cargo build"
                    }
                }
                "#,
        );
        let project = parsing::parse_project_file(&source).unwrap();
        let mut validator = Validator::new(files.root());
        let project_path = files
            .root()
            .subpath("service/")
            .unwrap()
            .validate()
            .unwrap();

        let project = validator.validate_project_definition(project, &project_path, &source);

        validator.ok().unwrap();

        let output = format!("{:#?}", project.unwrap())
            .replace(Utf8PathBuf::from(files.root()).as_str(), "[path]");

        insta::assert_snapshot!(output);
    }

    // Note: Most of the unhappy paths are covered by integration tests.
}
