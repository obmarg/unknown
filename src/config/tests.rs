use std::{fs::File, io::Read};

use camino::Utf8PathBuf;

use super::parsing::{parse_project_file, parse_task_file, parse_workspace_file};

#[test]
fn test_can_load_project_file() {
    let mut str_data = String::new();
    File::open("config-examples/project.kdl")
        .unwrap()
        .read_to_string(&mut str_data)
        .unwrap();

    insta::assert_debug_snapshot!(parse_project_file(
        &Utf8PathBuf::from("blah/project.kdl"),
        &dbg!(str_data)
    )
    .map_err(|e| miette::Report::new(e.0)))
}

#[test]
fn test_can_load_workspace_file() {
    let mut str_data = String::new();
    File::open("config-examples/workspace.kdl")
        .unwrap()
        .read_to_string(&mut str_data)
        .unwrap();

    insta::assert_debug_snapshot!(parse_workspace_file(
        &Utf8PathBuf::from("blah/workspace.kdl"),
        &str_data
    )
    .map_err(|e| miette::Report::new(e.0)))
}

#[test]
fn test_can_load_task_file() {
    let mut str_data = String::new();
    File::open("config-examples/task.kdl")
        .unwrap()
        .read_to_string(&mut str_data)
        .unwrap();

    insta::assert_debug_snapshot!(
        parse_task_file(&Utf8PathBuf::from("blah/task.kdl"), &str_data)
            .map_err(|e| miette::Report::new(e.0))
    )
}
