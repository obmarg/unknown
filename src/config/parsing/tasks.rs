use crate::config::{
    paths::{ConfigPath, ConfigPathValidationError, ValidPath},
    spanned::{SourceSpanExt, Spanned, WithSpan},
    validated, Glob, WorkspaceRoot,
};

use super::CollectResults;

#[derive(knuffel::Decode, Debug, Default)]
pub struct TaskBlock {
    #[knuffel(children(name = "import"), unwrap(argument))]
    pub(in crate::config) imports: Vec<ConfigPath>,

    // #[knuffel(children(name = "import_template"), unwrap(argument))]
    // template_imports: Vec<String>,
    #[knuffel(children(name = "task"))]
    pub(in crate::config) tasks: Vec<TaskDefinition>,
}

#[derive(thiserror::Error, miette::Diagnostic, Debug)]
#[error("A task file failed validation")]
pub enum TaskValidationError {
    InvalidPaths(#[related] Vec<ConfigPathValidationError>),
    InvalidTasks(#[related] Vec<TaskValidationError>),
    InvalidRequires(#[related] Vec<TaskValidationError>),
    MalformedRequire {
        span: miette::SourceSpan,
        message: String,
    },
}

#[derive(knuffel::Decode, Debug)]
pub struct TaskDefinition {
    #[knuffel(argument)]
    pub(super) name: String,

    #[knuffel(children(name = "command"), unwrap(argument))]
    pub(super) commands: Vec<String>,

    #[knuffel(children(name = "require"))]
    pub(super) requires: Vec<TaskRequires>,

    #[knuffel(children(name = "inputs"))]
    pub(super) input_blocks: Vec<InputBlock>,
}

impl From<InputBlock> for validated::InputBlock {
    fn from(value: InputBlock) -> Self {
        validated::InputBlock {
            paths: value.paths,
            env_vars: value.env_vars,
            commands: value.commands,
        }
    }
}

#[derive(knuffel::Decode, Debug)]
pub struct TaskRequires {
    #[knuffel(property)]
    task: Spanned<String>,

    #[knuffel(property(name = "in"))]
    target: Spanned<String>,
}

impl TaskRequires {
    pub fn parse(
        self,
        workspace_root: &WorkspaceRoot,
    ) -> Result<validated::TaskRequires, TaskValidationError> {
        use chumsky::Parser;
        use target_selector::ParsedSelector;

        let task_span = self.task.span.clone();

        let target = target_selector::parser()
            .parse(self.task.as_str())
            .map_err(|e| {
                let err = e.first().unwrap();
                let err_span = err.span();
                TaskValidationError::MalformedRequire {
                    span: task_span.subspan(err_span.start, err_span.len()),
                    message: err.to_string(),
                }
            })?;

        let target = match target {
            ParsedSelector::Project(inner, span) => validated::TargetSelector::Project(
                inner
                    .validate(workspace_root)
                    .map_err(|e| {
                        TaskValidationError::InvalidPaths(vec![ConfigPathValidationError::new(
                            e,
                            task_span.subspan(span.start, span.len()),
                        )])
                    })?
                    .with_span(task_span.subspan(span.start, span.len())),
            ),
            ParsedSelector::ProjectWithDependencies(inner, span) => {
                validated::TargetSelector::ProjectWithDependencies(
                    inner
                        .validate(workspace_root)
                        .map_err(|e| {
                            TaskValidationError::InvalidPaths(vec![ConfigPathValidationError::new(
                                e,
                                task_span.subspan(span.start, span.len()),
                            )])
                        })?
                        .with_span(task_span.subspan(span.start, span.len())),
                )
            }
            ParsedSelector::JustDependencies(inner, span) => {
                validated::TargetSelector::JustDependencies(
                    inner
                        .validate(workspace_root)
                        .map_err(|e| {
                            TaskValidationError::InvalidPaths(vec![ConfigPathValidationError::new(
                                e,
                                task_span.subspan(span.start, span.len()),
                            )])
                        })?
                        .with_span(task_span.subspan(span.start, span.len())),
                )
            }
        };

        Ok(validated::TaskRequires {
            task: self.task.into_inner(),
            target,
        })
    }
}

#[derive(knuffel::Decode, Debug)]
pub struct InputBlock {
    #[knuffel(children(name = "path"), unwrap(argument))]
    paths: Vec<Glob>,

    #[knuffel(children(name = "env_var"), unwrap(argument))]
    env_vars: Vec<String>,

    #[knuffel(children(name = "command"), unwrap(argument))]
    commands: Vec<String>,
}

mod target_selector {
    use std::ops::Range;

    use camino::Utf8PathBuf;
    use chumsky::prelude::*;
    use knuffel::{ast::Literal, decode::Kind, errors::DecodeError, traits::ErrorSpan};

    use crate::config::{paths::PathError, validated, TargetAnchor, TargetSelector, WorkspaceRoot};

    #[derive(Clone, Debug)]
    pub enum ParsedSelector {
        Project(ParsedAnchor, Range<usize>),
        ProjectWithDependencies(ParsedAnchor, Range<usize>),
        JustDependencies(ParsedAnchor, Range<usize>),
    }

    #[derive(Clone, Debug)]
    pub enum ParsedAnchor {
        CurrentProject,
        ProjectByName(String),
        ProjectByPath(Utf8PathBuf),
    }

    impl ParsedAnchor {
        pub fn validate(self, workspace_root: &WorkspaceRoot) -> Result<TargetAnchor, PathError> {
            Ok(match self {
                ParsedAnchor::CurrentProject => TargetAnchor::CurrentProject,
                ParsedAnchor::ProjectByName(name) => TargetAnchor::ProjectByName(name),
                ParsedAnchor::ProjectByPath(path) => {
                    TargetAnchor::ProjectByPath(workspace_root.subpath(path)?.validate()?)
                }
            })
        }
    }

    pub fn parser() -> impl chumsky::Parser<char, ParsedSelector, Error = Simple<char>> {
        let is_package_char = |c: &char| c.is_alphabetic() || *c == '_' || *c == '-';
        let is_path_char = |c: &char| true;

        let package_name = filter(is_package_char)
            .map(Some)
            .chain::<char, Vec<_>, _>(filter(is_package_char).repeated())
            .collect::<String>()
            .map(|name| ParsedAnchor::ProjectByName(name));

        let package_path = filter(|c| *c == '/')
            .map(Some)
            .chain::<char, Vec<_>, _>(filter(is_path_char).repeated())
            .collect::<String>()
            .from_str()
            .try_map(|val: Result<Utf8PathBuf, _>, span| {
                val.map_err(|e| Simple::custom(span, format!("Couldn't parse a path: {e}")))
            })
            .map(|name| ParsedAnchor::ProjectByPath(name));

        let anchor = choice::<_, Simple<char>>((
            text::keyword("self").to(ParsedAnchor::CurrentProject),
            package_path,
            package_name,
        ));

        choice::<_, Simple<char>>((
            text::keyword("...^")
                .ignore_then(anchor.clone())
                .map_with_span(ParsedSelector::JustDependencies),
            text::keyword("...")
                .ignore_then(anchor.clone())
                .map_with_span(ParsedSelector::ProjectWithDependencies),
            anchor.map_with_span(ParsedSelector::Project),
        ))
    }
}
