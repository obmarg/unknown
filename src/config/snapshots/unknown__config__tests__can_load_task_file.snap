---
source: src/config/tests.rs
expression: "parse_task_file(&Utf8PathBuf::from(\"blah/task.kdl\"),\n        &str_data).map_err(|e| miette::Report::new(e.0))"
---
Ok(
    TaskBlock {
        imports: [],
        tasks: [
            TaskDefinition {
                name: "xyz",
                commands: [
                    "cargo build",
                ],
                dependencies: [],
                input_blocks: [],
            },
        ],
    },
)
